packages/SwingSet/src/lib/capdata.js

```diff
@@ -30,15 +31,11 @@ export function insistCapData(capdata) {
  * @returns {string | undefined}
  */
 export function extractSingleSlot(data) {
-  const body = JSON.parse(data.body);
-  if (
-    body &&
-    typeof body === 'object' &&
-    body['@qclass'] === 'slot' &&
-    body.index === 0
-  ) {
-    if (data.slots.length === 1) {
-      return data.slots[0];
+  if (data.slots.length === 1) {
+    const encValue = kunser(data);
+    const slotValue = data.slots[0];
+    if (krefOf(encValue) === slotValue) {
```

Comment:
Seems like gilding the lily, since nobody is in a position to inject malformed capdata here.  Moreover, my read is that the new code is slightly stricter than what it replaces (due to the `krefOf(encValue) === slotValue` check), so I'm not sure what test you fear is being missed now that was present previously.

Reply:
> Seems like gilding the lily, since nobody is in a position to inject malformed capdata here.

The kernel [uses](https://github.com/Agoric/agoric-sdk/blob/68ecc09f4f57ed833acbd4dcc6de6140fae3f9e7/packages/SwingSet/src/kernel/kernel.js#L1030) this `extractSingleSlot()` when deciding how to route a delivery to a resolved promise, which means that the vat that did the `syscall.resolve` gets to put whatever it wants in the resolution capdata. Liveslots might not emit something malformed, but a compromised vat worker could, and anything coming in over comms has a capdata `.body` string produced by off-kernel senders, both of which are the kind of attackers I want to defend against. Comms [does the same thing](https://github.com/Agoric/agoric-sdk/blob/68ecc09f4f57ed833acbd4dcc6de6140fae3f9e7/packages/SwingSet/src/vats/comms/delivery.js#L266), for the same reason.

> Moreover, my read is that the new code is slightly stricter than what it replaces (due to the `krefOf(encValue) === slotValue` check), so I'm not sure what test you fear is being missed now that was present previously.

I think you're right, but I just thought of a different concern. Imagine that a vat or comms remote resolves a Promise to the serialized form of `{ getKref: 'haha', presence: Far(..) }`:
* the capdata will have a single `data.slots` for the presence's kref, but the top-level object will be a record, not a Far
* when someone sends a message to this promise, the kernel will call `extractSingleSlot()` to see if the resolution is an object
* `data.slots.length === 1`, `encValue = { getKref: 'haha', presence: Far(iface, { iface: () => iface, getKref: () => kref })}`
* we call `krefOf(encValue)`, which doesn't find `obj` in `fromMap`, but *does* see a truthy `obj.getKref` property, so it calls `obj.getKref()`, which throws with a TypeError
* that throws inside the kernel as it's trying to deliver a message, which probably causes a kernel panic

We really need this to be robust against adversarial capdata, even well-formed "normal" capdata.

Reply:
It sounds like we need `krefOf` to verify that the `getKref` property is not only present but actually a function, and if not simply return `null` like it would for any other not-just-a-single-presence-value value.